src/mrpro/operators/_FiniteDifferenceOp.py

```diff
@@ -0,0 +1,131 @@
+    for the copyright statement.
+
+    """
+
+    @staticmethod
+    def finite_difference_kernel(mode: str) -> torch.Tensor:
+        """Finite difference kernel.
+
+        Parameters
+        ----------
+        mode
+            String specifying kernel type
+
+        Returns
+        -------
+            Finite difference kernel
+
+        Raises
+        ------
+        ValueError
+            If mode is not central, forward, backward or doublecentral
+        """
+        if mode == 'doublecentral':
+            kernel = torch.tensor((-1, 0, 1))
+        elif mode == 'central':
+            kernel = torch.tensor((-1, 0, 1)) / 2
+        elif mode == 'forward':
+            kernel = torch.tensor((0, -1, 1))
+        elif mode == 'backward':
+            kernel = torch.tensor((-1, 1, 0))
+        else:
+            raise ValueError(f'mode should be one of (central, forward, backward, doublecentral), not {mode}')
+        return kernel
+
+    def __init__(
+        self,
+        dim: Sequence[int],
+        mode: Literal['doublecentral', 'central', 'forward', 'backward'] = 'doublecentral',
+        padding_mode: Literal['zeros', 'reflect', 'replicate', 'circular'] = 'zeros',
+    ) -> None:
+        """Finite difference operator.
+
+        Parameters
+        ----------
+        dim
+            Dimension along which finite differences are calculated.
+        mode
+            Type of finite difference operator
+        padding_mode
+            Padding to ensure output has the same size as the input
+        """
+        super().__init__()
+        self.dim = dim
+        self.padding_mode = padding_mode
+        self.kernel = self.finite_difference_kernel(mode)
+        self.adjoint_kernel = torch.flip(self.kernel, dims=(-1,))
+
+    def forward(self, x: torch.Tensor) -> tuple[torch.Tensor,]:
+        """Forward of finite differences.
+
+        Parameters
+        ----------
+        x
+            Input tensor.
+
+        Returns
+        -------
+            Finite differences of x along dim stacked along first dimension.
+        """
+        return (torch.stack([_filter_separable(x, (self.kernel,), axis=(d,)) for d in self.dim]),)
+
+    def adjoint(self, y: torch.Tensor) -> tuple[torch.Tensor,]:
+        """Adjoing of finite differences.
+
+        Parameters
+        ----------
+        y
+            Finite differences stacked along first dimension
+
+        Returns
+        -------
+            Adjoint finite differences
+
+        Raises
+        ------
+        ValueError
+            If the first dimension of y is to the same as the number of dimensions along which the finite differences
+            are calculated
+        """
+        if y.shape[0] != len(self.dim):
+            raise ValueError('Fist dimension of input tensor has to match the number of finite difference directions.')
+        return (
+            torch.sum(
```

Comment:
`sum(_filter_separable(yi, (self.adjoint_kernel,), axis=(dim,)) for dim,yi in zip(self.dim,y))`?

Reply:
Or what is the reason to first create the stacked intermediate tensor before summing over it?